igvm/drbd.py

```diff
@@ -267,44 +271,57 @@ def replicate_from_master(self, transaction=None):
             raise
 
     def wait_for_sync(self):
-        # Display a "nice" progress bar
-        show_progress = True
-        retry = 5
-        while show_progress:
-            lines = iter(self.hv.read_file('/proc/drbd').splitlines())
-            for line in lines:
-                line = line.decode()
-                if '{}: cs:'.format(self.get_device_minor()) in line:
-                    if 'ds:UpToDate/UpToDate' in line:
-                        show_progress = False
-                    try:
-                        next(lines)
-                        line = next(lines).decode().lstrip()
-                    except StopIteration:
-                        if retry:
-                            log.info(
-                                'No status yet, trying {} more time'
-                                .format(retry)
-                            )
-                            retry -= 1
-                            break
-                        log.warning(
-                            'Could not find progress bar, '
-                            'migrating without it!'
-                        )
-                        show_progress = False
-                    else:
-                        log.info(line)
-                    break
-            else:
-                # Exit the loop if status for current device can't be found
-                show_progress = False
-            sleep(1)
+        rep, size, _ = self._get_statistics()
+        last_n_synced = 0
+
+        with tqdm(
+            desc=self.vm_name,
+            total=size,
+            unit='iB',
+            unit_scale=True,
+            unit_divisor=1024,
+        ) as pbar:
+            while rep != 'Established' and rep is not None:
+                rep, _, oos = self._get_statistics()
+                n_synced = size - oos
 
-        # Progress bar does not determine if disks were synced.
-        # Wait for sync to be reported by DRBD.
+                pbar.update(n_synced - last_n_synced)
+                last_n_synced = n_synced
+                sleep(1)
+
+        # Wait for sync to be reported by DRBD in any case.
         self.hv.run('drbdsetup wait-sync {}'.format(self.get_device_minor()))
 
+    def _get_statistics(self) -> Tuple[
+        Optional[str],
+        Optional[int],
+        Optional[int],
+    ]:
+        """Returns a tuple containing statistics about the DRBD status."""
+        rep = None
+        size = None
+        oos = None
+
+        events = self.hv.run('drbdsetup events2 --now --statistics {}'.format(
```

Comment:
Does it work on Stretch HVs too?

Reply:
yes, tested a migration from stretch to buster and both hvs were showing the same output